themes/absimple/absimple.theme.sh

```diff
@@ -0,0 +1,384 @@
+}
+
+# Begin a segment on the right
+# Takes two arguments, background and foreground. Both can be omitted,
+# rendering default background/foreground.
+prompt_right_segment() {
+    local bg fg
+    declare -a codes
+
+    debug "Prompt right"
+    debug "Prompting $1 $2 $3"
+
+    # if commented out from kruton's original... I'm not clear
+    # if it did anything, but it messed up things like
+    # prompt_status - Erik 1/14/17
+
+    #    if [[ -z $1 || ( -z $2 && $2 != default ) ]]; then
+    codes=("${codes[@]}" $(text_effect reset))
+    #    fi
+    if [[ -n $1 ]]; then
+        bg=$(bg_color $1)
+        codes=("${codes[@]}" $bg)
+        debug "Added $bg as background to codes"
+    fi
+    if [[ -n $2 ]]; then
+        fg=$(fg_color $2)
+        codes=("${codes[@]}" $fg)
+        debug "Added $fg as foreground to codes"
+    fi
+
+    debug "Right Codes: "
+    # declare -p codes
+
+    # right always has a separator
+    # if [[ $CURRENT_RBG != NONE && $1 != $CURRENT_RBG ]]; then
+    #     $CURRENT_RBG=
+    # fi
+    declare -a intermediate2=($(fg_color $1) $(bg_color $CURRENT_RBG) )
+    # PRIGHT="$PRIGHT---"
+    debug "pre prompt " $(ansi_r intermediate2[@])
+    PRIGHT="$PRIGHT$(ansi_r intermediate2[@])$RIGHT_SEPARATOR"
+    debug "post prompt " $(ansi_r codes[@])
+    PRIGHT="$PRIGHT$(ansi_r codes[@]) "
+    # else
+    #     debug "no current BG, codes is $codes[@]"
+    #     PRIGHT="$PRIGHT$(ansi codes[@]) "
+    # fi
+    CURRENT_RBG=$1
+    [[ -n $3 ]] && PRIGHT="$PRIGHT$3"
+}
+
+######################################################################
+## Emacs prompt --- for dir tracking
+# stick the following in your .emacs if you use this:
+
+# (setq dirtrack-list '(".*DIR *\\([^ ]*\\) DIR" 1 nil))
+# (defun dirtrack-filter-out-pwd-prompt (string)
+#   "dirtrack-mode doesn't remove the PWD match from the prompt.  This does."
+#   ;; TODO: support dirtrack-mode's multiline regexp.
+#   (if (and (stringp string) (string-match (first dirtrack-list) string))
+#       (replace-match "" t t string 0)
+#     string))
+# (add-hook 'shell-mode-hook
+#           #'(lambda ()
+#               (dirtrack-mode 1)
+#               (add-hook 'comint-preoutput-filter-functions
+#                         'dirtrack-filter-out-pwd-prompt t t)))
+
+prompt_emacsdir() {
+    # no color or other setting... this will be deleted per above
+    PR="DIR \w DIR$PR"
+}
+
+######################################################################
+## Main prompt
+
+build_prompt() {
+    [[ ! -z ${AG_EMACS_DIR+x} ]] && prompt_emacsdir
+    prompt_status
+    [[ -z ${AG_NO_HIST+x} ]] && prompt_histdt
+    #[[ -z ${AG_NO_CONTEXT+x} ]] && prompt_context
+    prompt_virtualenv
+    prompt_dir
+    prompt_git
+    prompt_end
+}
+
+# from orig...
+# export PS1='$(ansi_single $(text_effect reset)) $(build_prompt) '
+# this doesn't work... new model: create a prompt via a PR variable and
+# use that.
+
+set_bash_prompt() {
+    RETVAL=$?
+    PR=""
```

Comment:
I think `PR` can also be made `local` to `set_bash_prompt`. Other functions touch the variable but are all called from `set_bash_prompt`. Could you also investigate whether the other variables around here, such as `RETVAL` and `PRIGHT`, may also be made `local`?

Reply:
i would never guessed this. done.